controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,279 @@
+}
+
+func (kc *KubernetesPodController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesPodController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesPodController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("%v: Error syncing Longhorn kubernetes pod %v: %v", controllerAgentName, key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("%v: Dropping Longhorn kubernetes pod %v out of the queue: %v", controllerAgentName, key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesPodController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", controllerAgentName, key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	pod, err := kc.pLister.Pods(namespace).Get(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Pod: %s", name)
+	}
+	nodeID := pod.Spec.NodeName
+	if err := kc.forceDeletePod(pod, nodeID, namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Because k8s never forcefully deletes pods of StatefullSet on a downed node,
+// the pods are stuck in terminating state forever and Longhorn volumes are not released.
+// We provide an option for users to help them automatically force delete terminating pods
+// of StatefulSet on the downed node. By force deleting, k8s will detach Longhorn volumes
+// and spin up replacement pods on a new node.
+//
+// Force delete a pod when all of the below conditions are meet:
+// 1. StatefulSetPodDeletePolicy is StatefulSetPodDeletePolicyForceDelete
+// 2. pod belongs to a StatefulSet
+// 3. node containing the pod is down
+// 4. the pod is terminating and the DeletionTimestamp has passed.
+// 5. pod has a PV with provisioner driver.longhorn.io
+func (kc *KubernetesPodController) forceDeletePod(pod *v1.Pod, nodeID string, namespace string) error {
+	deletionPolicy := types.StatefulSetPodDeletePolicyNever
+	if deletionSetting, err := kc.ds.GetSettingValueExisted(types.SettingNameStatefulSetPodDeletePolicy); err == nil {
+		deletionPolicy = types.StatefulSetPodDeletePolicy(deletionSetting)
+	}
+
+	if deletionPolicy == types.StatefulSetPodDeletePolicyNever {
+		return nil
+	}
+
+	if !isOwnedByStatefulSet(pod) {
+		return nil
+	}
+
+	isNodeDown, err := kc.ds.IsNodeDownOrDeleted(nodeID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to evaluate Node %v for pod %v in forceDeletePod", nodeID, pod.Name)
+	}
+	if !isNodeDown {
+		return nil
+	}
+
+	if pod.DeletionTimestamp == nil || pod.DeletionTimestamp.After(time.Now()) {
```

Comment:
Will Kubernetes delete the StatefulSet pod automatically?

Reply:
Kubernetes sets the `DeletionTimestamp`, the pod's status becomes `terminating`. But the pod is not deleted by Kubelet on the downed node and stuck in `terminating` forever.

Reply:
Does that mean user needs to delete the pod manually?

Reply:
No, users don't need to take any action if they allow Longhorn to force delete pods. The following will happen when a node (say `node-x`) goes down:

||| 
| :------------- | :----------: | 
| 0m:00s | `node-x`goes down and stops sending heartbeats to Kubernetes Node controller  |
| 0m:40s | Kubernetes Node controller reports `node-x` is `NotReady`.    |
| 5m:40s | Kubernetes Node controller starts evicting pods from `node-x` using graceful termination (set `DeletionTimestamp` and `deletionGracePeriodSeconds = 10s`)  |
| 5m:50s | Longhorn forces delete the pod of StatefulSet which uses Longhorn volume |

https://kubernetes.io/docs/concepts/architecture/nodes/#node-controller

I think that I might not consider the `deletionGracePeriodSeconds` in the current implementation. I will double check it.